LuGre friction model implementation (#97)
robotology/icub-firmware-models: sensorless_torque/README.md

that corresponds to the folder `icub-firmware\emBODY\eBcode\arch-arm\mbd\torque_estimator`

![folder](assets/Folder.png)

# LuGre parameters estimation

> An automatic parameter estimation tool is provided in this folder:
> 
> [param_estimator.m](param_estimator.m)

Reviewer: ```suggestion
An automatic parameter estimation tool is provided:
- [`param_estimator.m`](param_estimator.m)
```

The algorithm uses brute force in iterative steps, starting from a wide range of possible values and coarse quantization, and going on narrowing the range and increasing the resolution at each step.  
Since the number of parameters is still quite high, the algorithm estimates the LuGre parameters in 3 stages:

> 1.  Km (motor torque constant), Kw (viscous friction constant), Fc (Coulomb force constant)
> 2.  Fs (Stribeck force constant), Vth (Stribeck force velocity threshold)
> 3.  S0, S1 (LuGre hysteresis model stiffness and damping).

Reviewer: ```suggestion
1.  `Km` (motor torque constant), `Kw` (viscous friction constant), `Fc` (Coulomb force constant);
2.  `Fs` (Stribeck force constant), `Vth` (Stribeck force velocity threshold);
3.  `S0`, `S1` (LuGre hysteresis model stiffness and damping).
```

This formatting helps highlight variables within the text.

> The search space must be initially limited in some way, thus the Km torque constant, Kw viscous friction constant and Fc Coulomb force constant initial ranges are restricted to 1/5 to 5x the values measured in the ErgoCub new medium joint, covering in this way a wide range of present and future joint sizes.

Reviewer: ```suggestion
The search space must be initially limited in some way, thus the `Km` torque constant, `Kw` viscous friction constant and `Fc` Coulomb force constant initial ranges are restricted to 1/5x to 5x the values measured in the ergoCub new medium joint, covering in this way a wide range of present and future joint sizes.
```

## Data input format
The script requires an input with a 4xN matrix containing:

> 1.  time [s]
> 2.  measured current [A]
> 3.  measured velocity [rad/s], motor side
> 4.  measured joint torque [Nm], joint side.

Reviewer: ```suggestion
1.  time [s];
2.  measured current [A];
3.  measured velocity [rad/s], motor side;
4.  measured joint torque [Nm], joint side.
```

> The data source experiment should be in some way meaningful, i.e. with typical velocity and torque back and forth motion cycles, duration > 20 s, about 1 kHz sampling. 

Reviewer: ```suggestion
The data source experiment should be in some way meaningful: i.e., undergoing typical velocity and torque, back and forth motion cycles, time duration > 20 s, about 1 kHz sample rate. 
```

Example data file: [ergojoint_mid_calibration.mat](data/ergojoint_mid_calibration.mat)

![Example data plot](assets/DataExample.png)

The file path must be inserted in the script first lines, here:

> ![Input filepath](assets/filepath.png)

Reviewer: ```suggestion
![Input filepath](assets/FilePath.png)
```
